src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -557,12 +559,21 @@ public void MoveNext()
                     ExecutionContext.RunInternal(context, s_callback, this);
                 }
 
-                // In case this is a state machine box with a finalizer, suppress its finalization
-                // if it's now complete.  We only need the finalizer to run if the box is collected
-                // without having been completed.
-                if (IsCompleted && AsyncMethodBuilderCore.TrackAsyncMethodCompletion)
+                if (IsCompleted)
                 {
-                    GC.SuppressFinalize(this);
+                    // Clear out state now that the async method has completed.
+                    // This avoids keeping arbitrary state referenced by lifted locals
+                    // if this Task / state machine box is held onto.
+                    StateMachine = default;
+                    Context = default;
+
+                    // In case this is a state machine box with a finalizer, suppress its finalization
+                    // as it's now complete.  We only need the finalizer to run if the box is collected
+                    // without having been completed.
+                    if (AsyncMethodBuilderCore.TrackAsyncMethodCompletion)
```

Comment:
(Unrelated to this change) Would it make sense to move this into `DebugFinalizableAsyncStateMachineBox` so that this is not on the golden path? And also move creation of `new DebugFinalizableAsyncStateMachineBox` into its own non-inlineable method so that the DebugFinalizableAsyncStateMachineBox type is built only if you have the tracing on.

Reply:
> Would it make sense to move this into DebugFinalizableAsyncStateMachineBox so that this is not on the golden path?

Ideally, yes.  I'd looked at that previously and didn't find a great way to do it without introducing another virtual onto the golden path (and I preferred the extra branch to the virtual invocation), but several things have changed since then, so there might be a better way now.  I can look again.

> And also move creation of new DebugFinalizableAsyncStateMachineBox into its own non-inlineable method so that the DebugFinalizableAsyncStateMachineBox type is built only if you have the tracing on.

You mean at https://github.com/dotnet/coreclr/pull/20294/files/08e4e4f6f222919ed6d0f0430fb6922dc9973508#diff-68dc11e2c63ba209e0f903d3824d8464R488?  By "built", you're referring to JIT work?  Sure, I can move that.  I'll do it separately.

Reply:
> You mean at https://github.com/dotnet/coreclr/pull/20294/files/08e4e4f6f222919ed6d0f0430fb6922dc9973508#diff-68dc11e2c63ba209e0f903d3824d8464R488?

Yes, this one. The JIT triggers type building for all types directly referenced by the method even if they are on the cold path. It takes both time and memory -  each DebugFinalizableAsyncStateMachineBox type costs several hundred bytes at minimum.

Reply:
Makes sense.  I'll move it out separately.